tests/saitests/py3/sai_qos_tests.py

```diff
@@ -787,6 +796,134 @@ def runTest(self):
+                        inner_dscp = DEFAULT_DSCP
+                        exp_dscp = outer_dscp
+                    elif decap_mode == "pipe":
+                        outer_dscp = DEFAULT_DSCP
+                        inner_dscp = dscp
+                        exp_dscp = inner_dscp
+
+                    inner_pkt = simple_tcp_packet(ip_src=inner_src_port_ip,
+                                                  ip_dst=inner_dst_port_ip,
+                                                  ip_dscp=inner_dscp,
+                                                  ip_ecn=DEFAULT_ECN,
+                                                  ip_ttl=DEFAULT_TTL)
+
+                    inner_pkt.ttl -= 1
+
+                    outer_pkt = simple_ipv4ip_packet(eth_src=src_port_mac,
+                                                     eth_dst=pkt_dst_mac,
+                                                     ip_src=outer_src_port_ip,
+                                                     ip_dst=outer_dst_port_ip,
+                                                     ip_dscp=outer_dscp,
+                                                     ip_ecn=DEFAULT_ECN,
+                                                     inner_frame=inner_pkt[scapy.IP])
+
+                    inner_pkt.ttl += 1
+
+                    exp_pkt = simple_tcp_packet(ip_src=inner_src_port_ip,
+                                                ip_dst=inner_dst_port_ip,
+                                                ip_dscp=exp_dscp,
+                                                ip_ecn=DEFAULT_ECN,
+                                                ip_ttl=DEFAULT_TTL)
+
+                    exp_pkt = Mask(exp_pkt)
+                    exp_pkt.set_do_not_care_scapy(Ether, 'src')
+                    exp_pkt.set_do_not_care_scapy(Ether, 'dst')
+                    exp_pkt.set_do_not_care_scapy(IP, 'id')
+                    exp_pkt.set_do_not_care_scapy(IP, 'ttl')
+                    exp_pkt.set_do_not_care_scapy(IP, 'chksum')
+
+                    send_packet(self, src_port_id, outer_pkt, DEFAULT_PKT_COUNT)
+
+                    try:
+                        port_index, _ = verify_packet_any_port(self, exp_pkt, ports=upstream_ptf_ports, timeout=3)
+                    except AssertionError:
+                        cause_for_failure[0] = True
+                        cause_for_failure[1].append("Expected packet with DSCP {} was not received ".format(dscp) +
+                                                    "on any of the ports: {}".format(upstream_ptf_ports))
+
+                    # validate pg counters increment by the correct pkt num
+                    time.sleep(1)
+                    pg_cntrs = sai_thrift_read_pg_counters(self.src_client, port_list['src'][src_port_id])
+
+                    for i in range(0, PG_NUM):
+                        try:
+                            if i == pg:
+                                assert (pg_cntrs[pg] == pg_cntrs_base[pg] + DEFAULT_PKT_COUNT)
```

Comment:
There can be some background traffic, which may increate the PG counter. So I suggest adding some error tolerance here.

Reply:
ok sure